mlir/lib/Dialect/AIR/IR/AIRDialect.cpp

```diff
@@ -1108,6 +1109,390 @@ void WaitAllOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+                             strides))) {
+    return failure();
+  }
+
+  auto newOp = rewriter.create<air::DmaMemcpyNdOp>(
+      loc, op->getResultTypes(), op.getAsyncDependencies(), op.getDstMemref(),
+      op.getDstOffsets(), op.getDstSizes(), op.getDstStrides(), input_memref,
+      offsets, sizes, strides);
+
+  for (unsigned i = 0; i < op->getNumResults(); i++) {
+    op->getResult(i).replaceAllUsesWith(newOp->getResult(i));
+  }
+
+  rewriter.eraseOp(op);
+
+  return success();
+}
+
+static LogicalResult
+ComposeMemrefOpOnDmaMemcpyNdDst(DmaMemcpyNdOp op, PatternRewriter &rewriter) {
+
+  Value memref = op.getDstMemref();
+  if (!memref)
+    return failure();
+  auto loc = op->getLoc();
+  Value input_memref;
+  SmallVector<Value> offsets, sizes, strides;
+  offsets = op.getDstOffsets();
+  if (!offsets.empty())
+    return failure();
+  sizes = op.getDstSizes();
+  if (!sizes.empty())
+    return failure();
+  strides = op.getDstStrides();
+  if (!strides.empty())
+    return failure();
+
+  if (failed(ComposeMemrefOp(memref, rewriter, input_memref, offsets, sizes,
+                             strides))) {
+    return failure();
+  }
+
+  auto newOp = rewriter.create<air::DmaMemcpyNdOp>(
+      loc, op->getResultTypes(), op.getAsyncDependencies(), input_memref,
+      offsets, sizes, strides, op.getSrcMemref(), op.getSrcOffsets(),
+      op.getSrcSizes(), op.getSrcStrides());
+
+  for (unsigned i = 0; i < op->getNumResults(); i++) {
+    op->getResult(i).replaceAllUsesWith(newOp->getResult(i));
+  }
+
+  rewriter.eraseOp(op);
+
+  return success();
+}
+
+void DmaMemcpyNdOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+                                                MLIRContext *context) {
+  patterns.add(ComposeMemrefOpOnDmaMemcpyNdSrc);
+  patterns.add(ComposeMemrefOpOnDmaMemcpyNdDst);
+}
+
+//
+// Channel put op
+//
+
+static LogicalResult ComposeMemrefOpOnChannelPut(ChannelPutOp op,
+                                                 PatternRewriter &rewriter) {
+
+  Value memref = op.getMemref();
+  if (!memref)
+    return failure();
+  auto loc = op->getLoc();
+
+  // Init. memref type and offsets from memref's defining op's input type
+  Value input_memref;
+  SmallVector<Value> offsets, sizes, strides;
+  offsets = op.getOffsets();
+  if (!offsets.empty())
+    return failure();
+  sizes = op.getSizes();
+  if (!sizes.empty())
+    return failure();
+  strides = op.getStrides();
+  if (!strides.empty())
+    return failure();
+
+  if (failed(ComposeMemrefOp(memref, rewriter, input_memref, offsets, sizes,
+                             strides))) {
+    return failure();
+  }
+
+  auto newOp = rewriter.create<air::ChannelPutOp>(
+      loc, op->getResultTypes(), op.getAsyncDependencies(), op.getChanName(),
+      op.getIndices(), input_memref, offsets, sizes, strides);
+
+  for (unsigned i = 0; i < op->getNumResults(); i++) {
+    op->getResult(i).replaceAllUsesWith(newOp->getResult(i));
+  }
```

Comment:
Use the rewriter instead of direct IR modification:
```suggestion
  rewriter.replaceOp(op, newOp);
```

Reply:
Updated.